From 3c97bd85f09331b7881405de2f6eef2e9b438934 Mon Sep 17 00:00:00 2001
From: Pavel Snajdr <snajpa@snajpa.net>
Date: Wed, 14 Nov 2018 22:04:56 +0100
Subject: [PATCH] Allow 32 bit memory control IDs

---
 include/linux/memcontrol.h |  5 +++++
 init/Kconfig               | 11 +++++++++++
 2 files changed, 16 insertions(+)

diff --git a/include/linux/memcontrol.h b/include/linux/memcontrol.h
index 652f602167df..363102d2d8fd 100644
--- a/include/linux/memcontrol.h
+++ b/include/linux/memcontrol.h
@@ -73,8 +73,13 @@ struct mem_cgroup_reclaim_cookie {
 
 #ifdef CONFIG_MEMCG
 
+#if defined(__x86_64__) && defined(CONFIG_MEMCG_32BIT_IDS)
+#define MEM_CGROUP_ID_SHIFT	32
+#define MEM_CGROUP_ID_MAX	INT_MAX
+#else
 #define MEM_CGROUP_ID_SHIFT	16
 #define MEM_CGROUP_ID_MAX	USHRT_MAX
+#endif
 
 struct mem_cgroup_id {
 	int id;
diff --git a/init/Kconfig b/init/Kconfig
index 1e234e2f1cba..2dc7250d0efc 100644
--- a/init/Kconfig
+++ b/init/Kconfig
@@ -713,6 +713,17 @@ config MEMCG_KMEM
 	depends on MEMCG && !SLOB
 	default y
 
+config MEMCG_32BIT_IDS
+	bool "Use 32 bit IDs"
+	depends on MEMCG && X86_64
+	default n
+	help
+	  Extends the default limit of max 65536 memory cgroups to
+	  2147483647. This may cause an increase of bucket_order in
+	  mm/workingset.c, having TODO consequences.
+	  See https://lore.kernel.org/patchwork/patch/690171/ for original
+	  reasoning behind 16 bit limit.
+
 config BLK_CGROUP
 	bool "IO controller"
 	depends on BLOCK
-- 
2.15.0

