diff --git a/fs/nfsd/auth.c b/fs/nfsd/auth.c
index fdf2aad73470..80ee577dc8d4 100644
--- a/fs/nfsd/auth.c
+++ b/fs/nfsd/auth.c
@@ -76,7 +76,13 @@ int nfsd_setuser(struct svc_rqst *rqstp, struct svc_export *exp)
 	set_groups(new, gi);
 	put_group_info(gi);
 
-	if (!uid_eq(new->fsuid, GLOBAL_ROOT_UID))
+#if 0
+	printk("nfsd: auth: fsuid=%u, root_uid=%u\n",
+			__kuid_val(new->fsuid), __kuid_val(exp->ex_root_uid));
+#endif
+
+	if (!uid_eq(new->fsuid, GLOBAL_ROOT_UID)
+	    && !uid_eq(new->fsuid, exp->ex_root_uid))
 		new->cap_effective = cap_drop_nfsd_set(new->cap_effective);
 	else
 		new->cap_effective = cap_raise_nfsd_set(new->cap_effective,
diff --git a/fs/nfsd/export.c b/fs/nfsd/export.c
index a1143f7c2201..6965350ea114 100644
--- a/fs/nfsd/export.c
+++ b/fs/nfsd/export.c
@@ -517,6 +517,7 @@ static int svc_export_parse(struct cache_detail *cd, char *mesg, int mlen)
 	struct auth_domain *dom = NULL;
 	struct svc_export exp = {}, *expp;
 	int an_int;
+	unsigned int an_uint;
 
 	if (mesg[mlen-1] != '\n')
 		return -EINVAL;
@@ -584,6 +585,15 @@ static int svc_export_parse(struct cache_detail *cd, char *mesg, int mlen)
 			goto out3;
 		exp.ex_fsid = an_int;
 
+		/* root_uid */
+		err = get_uint(&mesg, &an_uint);
+		if (err)
+			goto out3;
+		exp.ex_root_uid = KUIDT_INIT(an_uint);
+#if 0
+		printk("nfsd: got root_uid=%u\n", an_uint);
+#endif
+
 		while ((len = qword_get(&mesg, buf, PAGE_SIZE)) > 0) {
 			if (strcmp(buf, "fsloc") == 0)
 				err = fsloc_parse(&mesg, buf, &exp.ex_fslocs);
@@ -723,6 +733,7 @@ static void export_update(struct cache_head *cnew, struct cache_head *citem)
 	new->ex_anon_uid = item->ex_anon_uid;
 	new->ex_anon_gid = item->ex_anon_gid;
 	new->ex_fsid = item->ex_fsid;
+	new->ex_root_uid = item->ex_root_uid;
 	new->ex_devid_map = item->ex_devid_map;
 	item->ex_devid_map = NULL;
 	new->ex_uuid = item->ex_uuid;
diff --git a/fs/nfsd/export.h b/fs/nfsd/export.h
index c8b74126ddaa..8ae14e5b2533 100644
--- a/fs/nfsd/export.h
+++ b/fs/nfsd/export.h
@@ -61,6 +61,7 @@ struct svc_export {
 	u32			ex_layout_types;
 	struct nfsd4_deviceid_map *ex_devid_map;
 	struct cache_detail	*cd;
+	kuid_t                   ex_root_uid;
 };
 
 /* an "export key" (expkey) maps a filehandlefragement to an
diff --git a/include/uapi/linux/nfsd/export.h b/include/uapi/linux/nfsd/export.h
index 2124ba904779..b1071c2bdfc0 100644
--- a/include/uapi/linux/nfsd/export.h
+++ b/include/uapi/linux/nfsd/export.h
@@ -53,6 +53,7 @@
  */
 #define	NFSEXP_V4ROOT		0x10000
 #define NFSEXP_PNFS		0x20000
+#define NFSEXP_ROOTUID          0x40000
 
 /* All flags that we claim to support.  (Note we don't support NOACL.) */
 #define NFSEXP_ALLFLAGS		0x3FEFF
