diff --git a/kernel/sched/core.c b/kernel/sched/core.c
index 10ca5868f62c..6826a9b6bbc9 100644
--- a/kernel/sched/core.c
+++ b/kernel/sched/core.c
@@ -8331,7 +8331,8 @@ long sched_getaffinity(pid_t pid, struct cpumask *mask)
 		goto out_unlock;
 
 	raw_spin_lock_irqsave(&p->pi_lock, flags);
-	cpumask_and(mask, &p->cpus_mask, cpu_active_mask);
+	if (!fake_affinity_cpumask(p, mask))
+		cpumask_and(mask, &p->cpus_mask, cpu_active_mask);
 	raw_spin_unlock_irqrestore(&p->pi_lock, flags);
 
 out_unlock:
diff --git a/kernel/vpsadminos.c b/kernel/vpsadminos.c
index 8481e391720e..a4ac436db47f 100644
--- a/kernel/vpsadminos.c
+++ b/kernel/vpsadminos.c
@@ -509,3 +509,23 @@ u64 fake_cputime_readout_idle(u64 timestamp, struct task_struct *p)
 	return (timestamp * cpus) - user - system;
 }
 
+#include <linux/utsname.h>
+#include <linux/vpsadminos-livepatch.h>
+#include "kpatch-macros.h"
+char old_uname[65];
+char new_uname[65];
+
+static int patch(patch_object *obj)
+{
+	scnprintf(new_uname, 64, "%s.%s", LIVEPATCH_ORIG_KERNEL_VERSION,
+	    LIVEPATCH_NAME);
+	scnprintf(old_uname, 64, "%s", init_uts_ns.name.release);
+	scnprintf(init_uts_ns.name.release, 64, "%s", new_uname);
+	return 0;
+}
+KPATCH_PRE_PATCH_CALLBACK(patch);
+static void unpatch(patch_object *obj)
+{
+	scnprintf(init_uts_ns.name.release, 64, "%s", old_uname);
+}
+KPATCH_POST_UNPATCH_CALLBACK(unpatch);
