diff --git a/kernel/printk/printk.c b/kernel/printk/printk.c
index f6e00c86b5c7..108f01c422c6 100644
--- a/kernel/printk/printk.c
+++ b/kernel/printk/printk.c
@@ -625,7 +625,7 @@ int devkmsg_emit(struct syslog_namespace *ns, int facility, int level,
 		va_list args2;
 		va_copy(args2, args);
 		va_start(args2, fmt);
-		vprintk_store_ns(&init_syslog_ns, facility, level, NULL, fmt, args2);
+		vprintk_emit_ns(&init_syslog_ns, facility, level, NULL, fmt, args2);
 		va_end(args2);
 	}
 
@@ -3567,8 +3567,12 @@ static void __wake_up_klogd(struct syslog_namespace *ns, int val)
 	 */
 	if (wq_has_sleeper(&ns->log_wait) || /* LMM(__wake_up_klogd:A) */
 	    (val & PRINTK_PENDING_OUTPUT)) {
-		this_cpu_or(printk_pending, val);
-		irq_work_queue(this_cpu_ptr(&wake_up_klogd_work));
+		if (ns == &init_syslog_ns) {
+			this_cpu_or(printk_pending, val);
+			irq_work_queue(this_cpu_ptr(&wake_up_klogd_work));
+		} else
+			if (val & PRINTK_PENDING_WAKEUP)
+				wake_up_interruptible(&ns->log_wait);
 	}
 	preempt_enable();
 }
diff --git a/kernel/vpsadminos.c b/kernel/vpsadminos.c
index 4350b6d01b90..d804df909ad2 100644
--- a/kernel/vpsadminos.c
+++ b/kernel/vpsadminos.c
@@ -10,6 +10,26 @@
 #include <asm/page.h>
 #include <linux/sched/cputime.h>
 #include "sched/sched.h"
+#include <linux/utsname.h>
+#include <linux/vpsadminos-livepatch.h>
+#include "kpatch-macros.h"
+char old_uname[65];
+char new_uname[65];
+
+static int patch(patch_object *obj)
+{
+	scnprintf(new_uname, 64, "%s.%s", LIVEPATCH_ORIG_KERNEL_VERSION,
+	    LIVEPATCH_NAME);
+	scnprintf(old_uname, 64, "%s", init_uts_ns.name.release);
+	scnprintf(init_uts_ns.name.release, 64, "%s", new_uname);
+	return 0;
+}
+KPATCH_PRE_PATCH_CALLBACK(patch);
+static void unpatch(patch_object *obj)
+{
+	scnprintf(init_uts_ns.name.release, 64, "%s", old_uname);
+}
+KPATCH_POST_UNPATCH_CALLBACK(unpatch);
 
 static int __init vpsadminos_init(void)
 {
