diff --git a/fs/coredump.c b/fs/coredump.c
index 778679075040..06564dc309f7 100644
--- a/fs/coredump.c
+++ b/fs/coredump.c
@@ -994,7 +994,7 @@ static int proc_dostring_coredump(struct ctl_table *table, int write,
 	struct user_namespace *ns = current_1stlvl_user_ns();
 	struct ctl_table tmp;
 
-	if (write && !ns_capable(ns, CAP_SYS_ADMIN))
+	if (write && !ns_capable(&init_user_ns, CAP_SYS_ADMIN))
 		return -EPERM;
 
 	tmp.data = ns->core_pattern;
diff --git a/kernel/vpsadminos.c b/kernel/vpsadminos.c
index 075cf864cb63..460b11b22535 100644
--- a/kernel/vpsadminos.c
+++ b/kernel/vpsadminos.c
@@ -10,6 +10,26 @@
 #include <asm/page.h>
 #include <linux/sched/cputime.h>
 #include "sched/sched.h"
+#include <linux/utsname.h>
+#include <linux/vpsadminos-livepatch.h>
+#include "kpatch-macros.h"
+char old_uname[65];
+char new_uname[65];
+
+static int patch(patch_object *obj)
+{
+	scnprintf(new_uname, 64, "%s.%s", LIVEPATCH_ORIG_KERNEL_VERSION,
+	    LIVEPATCH_NAME);
+	scnprintf(old_uname, 64, "%s", init_uts_ns.name.release);
+	scnprintf(init_uts_ns.name.release, 64, "%s", new_uname);
+	return 0;
+}
+KPATCH_PRE_PATCH_CALLBACK(patch);
+static void unpatch(patch_object *obj)
+{
+	scnprintf(init_uts_ns.name.release, 64, "%s", old_uname);
+}
+KPATCH_POST_UNPATCH_CALLBACK(unpatch);
 
 struct proc_dir_entry *proc_vpsadminos;
 
