diff --git a/kernel/sched/core.c b/kernel/sched/core.c
index 75267659648b..b98691bef1e7 100644
--- a/kernel/sched/core.c
+++ b/kernel/sched/core.c
@@ -9079,11 +9079,13 @@ static u64 cpu_shares_read_u64(struct cgroup_subsys_state *css,
 
 LIST_HEAD(cgns_avenrun_list);
 raw_spinlock_t cgns_avenrun_lock = __RAW_SPIN_LOCK_UNLOCKED(&cgns_avenrun_lock);
+static struct task_struct *cgns_avenrund_task;
+
+int cgns_avenrund_start(void);
+void cgns_avenrund_stop(void);
 
 void cgroup_ns_track_loadavg(struct cgroup_namespace *ns)
 {
-	unsigned long flags;
-
 	if (ns == &init_cgroup_ns)
 		return;
 
@@ -9093,15 +9095,17 @@ void cgroup_ns_track_loadavg(struct cgroup_namespace *ns)
 	if (ns->loadavg_virt_enabled)
 		return;
 
-	raw_spin_lock_irqsave(&cgns_avenrun_lock, flags);
+	cgns_avenrund_start();
+
+	raw_spin_lock(&cgns_avenrun_lock);
 	list_add(&ns->cgns_avenrun_list, &cgns_avenrun_list);
 	ns->loadavg_virt_enabled = 1;
-	raw_spin_unlock_irqrestore(&cgns_avenrun_lock, flags);
+	raw_spin_unlock(&cgns_avenrun_lock);
 }
 
 void cgroup_ns_untrack_loadavg(struct cgroup_namespace *ns)
 {
-	unsigned long flags;
+	bool stop_avenrund = false;
 
 	if (ns == &init_cgroup_ns)
 		return;
@@ -9109,15 +9113,20 @@ void cgroup_ns_untrack_loadavg(struct cgroup_namespace *ns)
 	if (!ns->loadavg_virt_enabled)
 		return;
 
-	raw_spin_lock_irqsave(&cgns_avenrun_lock, flags);
+	raw_spin_lock(&cgns_avenrun_lock);
 	list_del_init(&ns->cgns_avenrun_list);
 	ns->loadavg_virt_enabled = 0;
-	raw_spin_unlock_irqrestore(&cgns_avenrun_lock, flags);
+	if (list_empty(&cgns_avenrun_list))
+		stop_avenrund = true;
+	raw_spin_unlock(&cgns_avenrun_lock);
+
+	if (stop_avenrund)
+		cgns_avenrund_stop();
 }
 
 unsigned long cgns_nr_running(struct cgroup_namespace *ns)
 {
-	unsigned long flags, nr_active = 0;
+	unsigned long nr_active = 0;
 	struct task_group *tg;
 	int i;
 
@@ -9135,9 +9144,9 @@ unsigned long cgns_nr_running(struct cgroup_namespace *ns)
 
 	rcu_read_unlock();
 
-	raw_spin_lock_irqsave(&ns->cgns_avenrun_lock, flags);
+	raw_spin_lock(&ns->cgns_avenrun_lock);
 	nr_active += ns->nr_uninterruptible;
-	raw_spin_unlock_irqrestore(&ns->cgns_avenrun_lock, flags);
+	raw_spin_unlock(&ns->cgns_avenrun_lock);
 
 	return nr_active;
 }
@@ -9213,7 +9222,6 @@ unsigned long cgroup_ns_nr_threads(struct task_struct *p)
 void inc_cgns_nr_uninterruptible(struct task_struct *p)
 {
 	struct cgroup_namespace *ns;
-	unsigned long flags;
 
 	if (p->nsproxy && p->nsproxy->cgroup_ns)
 		ns = p->nsproxy->cgroup_ns;
@@ -9227,9 +9235,9 @@ void inc_cgns_nr_uninterruptible(struct task_struct *p)
 		ns = ns->parent;
 
 	if (ns->loadavg_virt_enabled) {
-		raw_spin_lock_irqsave(&ns->cgns_avenrun_lock, flags);
+		raw_spin_lock(&ns->cgns_avenrun_lock);
 		ns->nr_uninterruptible++;
-		raw_spin_unlock_irqrestore(&ns->cgns_avenrun_lock, flags);
+		raw_spin_unlock(&ns->cgns_avenrun_lock);
 		p->sched_contributed_to_load = ns;
 	}
 }
@@ -9244,39 +9252,37 @@ void dec_cgns_nr_uninterruptible(struct task_struct *p)
 		return;
 	BUG_ON(!ns->loadavg_virt_enabled);
 
-	raw_spin_lock_irqsave(&ns->cgns_avenrun_lock, flags);
+	raw_spin_lock(&ns->cgns_avenrun_lock);
 	ns->nr_uninterruptible--;
-	raw_spin_unlock_irqrestore(&ns->cgns_avenrun_lock, flags);
+	raw_spin_unlock(&ns->cgns_avenrun_lock);
 	p->sched_contributed_to_load = NULL;
 }
 
 void cgns_calc_avenrun(void)
 {
 	struct cgroup_namespace *ns;
-	unsigned long flags, flags2, nr_active;
+	unsigned long nr_active;
 
-	raw_spin_lock_irqsave(&cgns_avenrun_lock, flags);
+	raw_spin_lock(&cgns_avenrun_lock);
 	list_for_each_entry(ns, &cgns_avenrun_list, cgns_avenrun_list) {
 		nr_active = cgns_nr_running(ns);
 		nr_active = nr_active > 0 ? nr_active * FIXED_1 : 0;
-		raw_spin_lock_irqsave(&ns->cgns_avenrun_lock, flags2);
+		raw_spin_lock(&ns->cgns_avenrun_lock);
 		ns->avenrun[0] = calc_load(ns->avenrun[0], EXP_1, nr_active);
 		ns->avenrun[1] = calc_load(ns->avenrun[1], EXP_5, nr_active);
 		ns->avenrun[2] = calc_load(ns->avenrun[2], EXP_15, nr_active);
-		raw_spin_unlock_irqrestore(&ns->cgns_avenrun_lock, flags2);
+		raw_spin_unlock(&ns->cgns_avenrun_lock);
 	}
-	raw_spin_unlock_irqrestore(&cgns_avenrun_lock, flags);
+	raw_spin_unlock(&cgns_avenrun_lock);
 }
 
 void get_avenrun_fake_ns(struct cgroup_namespace *ns, unsigned long *loads, unsigned long offset, int shift)
 {
-	unsigned long flags;
-
-	raw_spin_lock_irqsave(&ns->cgns_avenrun_lock, flags);
+	raw_spin_lock(&ns->cgns_avenrun_lock);
 	loads[0] = (ns->avenrun[0] + offset) << shift;
 	loads[1] = (ns->avenrun[1] + offset) << shift;
 	loads[2] = (ns->avenrun[2] + offset) << shift;
-	raw_spin_unlock_irqrestore(&ns->cgns_avenrun_lock, flags);
+	raw_spin_unlock(&ns->cgns_avenrun_lock);
 }
 
 int get_avenrun_fake(struct task_struct *p, unsigned long *loads, unsigned long offset, int shift)
@@ -9299,14 +9305,14 @@ int get_avenrun_fake(struct task_struct *p, unsigned long *loads, unsigned long
 
 int virt_loadavg_proc_show(struct seq_file *m, void *v)
 {
-	unsigned long flags, avnrun[3];
+	unsigned long avnrun[3];
 	int nr_r; uint nr_t;
 	struct cgroup_namespace *ns;
 
 	if (!capable(CAP_SYS_ADMIN))
 		return -EPERM;
 
-	raw_spin_lock_irqsave(&cgns_avenrun_lock, flags);
+	raw_spin_lock(&cgns_avenrun_lock);
 	list_for_each_entry(ns, &cgns_avenrun_list, cgns_avenrun_list) {
 		get_avenrun_fake_ns(ns, avnrun, FIXED_1/200, 0);
 		nr_r = cgns_nr_running(ns);
@@ -9319,10 +9325,45 @@ int virt_loadavg_proc_show(struct seq_file *m, void *v)
 			LOAD_INT(avnrun[2]), LOAD_FRAC(avnrun[2]),
 			nr_r, nr_t);
 	}
-	raw_spin_unlock_irqrestore(&cgns_avenrun_lock, flags);
+	raw_spin_unlock(&cgns_avenrun_lock);
+	return 0;
+}
+
+static int cgns_avenrund(void *data)
+{
+	unsigned long start, delay;
+	while (!kthread_should_stop()) {
+		start = jiffies;
+		cgns_calc_avenrun();
+		delay = LOAD_FREQ - (jiffies - start);
+		schedule_timeout_interruptible(delay);
+	}
+
 	return 0;
 }
 
+int cgns_avenrund_start(void)
+{
+	if (cgns_avenrund_task)
+		return 0;
+
+	cgns_avenrund_task = kthread_run(cgns_avenrund, NULL, "cgns_avenrund");
+	if (IS_ERR(cgns_avenrund_task))
+		return PTR_ERR(cgns_avenrund_task);
+
+	pr_info("cgroup namespace loadavg tracking started\n");
+	return 0;
+}
+
+void cgns_avenrund_stop(void)
+{
+	if (cgns_avenrund_task)
+		kthread_stop(cgns_avenrund_task);
+	cgns_avenrund_task = NULL;
+	pr_info("cgroup namespace loadavg tracking stopped\n");
+}
+
+
 #ifdef CONFIG_CFS_BANDWIDTH
 static DEFINE_MUTEX(cfs_constraints_mutex);
 
diff --git a/kernel/time/tick-common.c b/kernel/time/tick-common.c
index 32ea9c06a120..a47bcf71defc 100644
--- a/kernel/time/tick-common.c
+++ b/kernel/time/tick-common.c
@@ -85,8 +85,6 @@ int tick_is_oneshot_available(void)
  */
 static void tick_periodic(int cpu)
 {
-	bool calc_load = false;
-
 	if (READ_ONCE(tick_do_timer_cpu) == cpu) {
 		raw_spin_lock(&jiffies_lock);
 		write_seqcount_begin(&jiffies_seq);
@@ -94,11 +92,9 @@ static void tick_periodic(int cpu)
 		/* Keep track of the next tick event */
 		tick_next_period = ktime_add_ns(tick_next_period, TICK_NSEC);
 
-		calc_load = do_timer(1);
+		do_timer(1);
 		write_seqcount_end(&jiffies_seq);
 		raw_spin_unlock(&jiffies_lock);
-		if (calc_load)
-			cgns_calc_avenrun();
 		update_wall_time();
 	}
 
diff --git a/kernel/time/tick-legacy.c b/kernel/time/tick-legacy.c
index 30102776e778..af225b32f5b3 100644
--- a/kernel/time/tick-legacy.c
+++ b/kernel/time/tick-legacy.c
@@ -24,16 +24,12 @@
  */
 void legacy_timer_tick(unsigned long ticks)
 {
-	bool calc_load = false;
-
 	if (ticks) {
 		raw_spin_lock(&jiffies_lock);
 		write_seqcount_begin(&jiffies_seq);
-		calc_load = do_timer(ticks);
+		do_timer(ticks);
 		write_seqcount_end(&jiffies_seq);
 		raw_spin_unlock(&jiffies_lock);
-		if (calc_load)
-			cgns_calc_avenrun();
 		update_wall_time();
 	}
 	update_process_times(user_mode(get_irq_regs()));
diff --git a/kernel/time/tick-sched.c b/kernel/time/tick-sched.c
index f54db7927af1..753a184c7090 100644
--- a/kernel/time/tick-sched.c
+++ b/kernel/time/tick-sched.c
@@ -58,7 +58,6 @@ static void tick_do_update_jiffies64(ktime_t now)
 {
 	unsigned long ticks = 1;
 	ktime_t delta, nextp;
-	bool calc_load = false;
 
 	/*
 	 * 64-bit can do a quick check without holding the jiffies lock and
@@ -144,11 +143,9 @@ static void tick_do_update_jiffies64(ktime_t now)
 	 */
 	write_seqcount_end(&jiffies_seq);
 
-	calc_load = calc_global_load();
+	calc_global_load();
 
 	raw_spin_unlock(&jiffies_lock);
-	if (calc_load)
-		cgns_calc_avenrun();
 	update_wall_time();
 }
 
diff --git a/kernel/time/timekeeping.c b/kernel/time/timekeeping.c
index 7666ab49cdbf..5391e4167d60 100644
--- a/kernel/time/timekeeping.c
+++ b/kernel/time/timekeeping.c
@@ -2415,10 +2415,10 @@ EXPORT_SYMBOL(ktime_get_coarse_ts64);
 /*
  * Must hold jiffies_lock
  */
-bool do_timer(unsigned long ticks)
+void do_timer(unsigned long ticks)
 {
 	jiffies_64 += ticks;
-	return calc_global_load();
+	calc_global_load();
 }
 
 /**
diff --git a/kernel/time/timekeeping.h b/kernel/time/timekeeping.h
index 4cab7fa8f970..543beba096c7 100644
--- a/kernel/time/timekeeping.h
+++ b/kernel/time/timekeeping.h
@@ -23,7 +23,7 @@ static inline void sched_clock_resume(void) { }
 #endif
 
 extern void update_process_times(int user);
-extern bool do_timer(unsigned long ticks);
+extern void do_timer(unsigned long ticks);
 extern void update_wall_time(void);
 
 extern raw_spinlock_t jiffies_lock;
diff --git a/kernel/vpsadminos.c b/kernel/vpsadminos.c
index 9369f91f9920..f9db0589ea2d 100644
--- a/kernel/vpsadminos.c
+++ b/kernel/vpsadminos.c
@@ -10,6 +10,31 @@
 #include <linux/cgroup.h>
 #include <linux/sched/cputime.h>
 #include "sched/sched.h"
+#include <linux/utsname.h>
+#include <linux/vpsadminos-livepatch.h>
+#include "kpatch-macros.h"
+char old_uname[65];
+char new_uname[65];
+
+extern int cgns_avenrund_start(void);
+extern void cgns_avenrund_stop(void);
+
+static int patch(patch_object *obj)
+{
+	cgns_avenrund_start();
+	scnprintf(new_uname, 64, "%s.%s", LIVEPATCH_ORIG_KERNEL_VERSION,
+	    LIVEPATCH_NAME);
+	scnprintf(old_uname, 64, "%s", init_uts_ns.name.release);
+	scnprintf(init_uts_ns.name.release, 64, "%s", new_uname);
+	return 0;
+}
+KPATCH_PRE_PATCH_CALLBACK(patch);
+static void unpatch(patch_object *obj)
+{
+	scnprintf(init_uts_ns.name.release, 64, "%s", old_uname);
+	cgns_avenrund_stop();
+}
+KPATCH_POST_UNPATCH_CALLBACK(unpatch);
 
 struct proc_dir_entry *proc_vpsadminos;
 
